wavekit doc (#484)
risingwavelabs/risingwave-docs: docs/manage/meta-backup.md

---
id: meta-backup
title: Back up and restore meta service
description: Back up and restore meta service
slug: /meta-backup
---

This guide introduces how to back up meta service data and restore from a backup.

## Configure system variables

A meta snapshot is a backup of meta service's data at a specific point in time. Meta snapshots are persisted in S3-compatible storage.

Here's an example of how to specify target storage and set `storage_url` and `storage_directory`:

```
[backup]
storage_url = "s3://[bucket]"
storage_directory = "backup"

> ```

Reviewer: From a user perspective, this is confusing since I don't know where to set these configs, especially when trying the pre-built binary.

In general, do we have a doc to guide user how to and where to set configs?

Meta snapshot is created by meta service.

Here's an example of how to create a new meta snapshot with `risectl`:

```
risectl meta backup-meta
```

`risectl` is included in the pre-built RisingWave binary. It can also be built from source. For details, see [Run RisingWave locally](/deploy/risingwave-local.md).

## View existing meta snapshots

The following SQL command lists existing meta snapshots:

```sql
SELECT meta_snapshot_id FROM rw_catalog.rw_meta_snapshot;
```

Example output:

```
 meta_snapshot_id 
------------------
                3
                4
```

## Delete a meta snapshot

Here's an example of how to delete a meta snapshot with `risectl`:

```
risectl meta delete-meta-snapshots [snapshot_ids]
```

## Restore from a meta snapshot

Use the following steps to restore from a meta snapshot.

1. Shut down the meta service.
    :::note
    This step is especially important because the meta backup and recovery process does not replicate SST files. It is not permitted for multiple clusters to run with the same SSTs set at any time, as this can corrupt the SST files.
    :::
2. Create an empty meta store.
3. Restore the meta snapshot to the new meta store.

    ```

>     backup-restore \

Reviewer: @zwang28 Should we make `backup-restore` part of `risectl` so user doesn't need to build another stand-alone binary?

    --meta-store-type etcd \
    --meta-snapshot-id [snapshot_id] \
    --etcd-endpoints [etcd_endpoints] \
    --storage-url [storage_url]
    ```
    `backup-restore` reads snapshot data from snapshot storage and writes them to `etcd`. 
    `backup-restore` is not included in the pre-built risingwave binary. Please build it from source by compiling the `risingwave_backup_cmd` package.
4. Configure meta service to use the new meta store.

> ## Access historical data backed up by meta snapshot

Reviewer: FYI, this part won't work when running risingwave pre-built binary or running with default backup config (with `storage_url` to be memory). @zwang28 and I are discussing how to deal with it.

---
id: meta-backup
title: Back up and restore meta service
description: Back up and restore meta service
slug: /meta-backup
---

This guide introduces how to back up meta service data and restore from a backup.

## Configure system variables

A meta snapshot is a backup of meta service's data at a specific point in time. Meta snapshots are persisted in S3-compatible storage.

Here's an example of how to specify target storage and set `storage_url` and `storage_directory`:

```
[backup]
storage_url = "s3://[bucket]"
storage_directory = "backup"
```

Typically, `storage_url` and `storage_directory` should not be changed after initializing the cluster. If they are changed, all meta snapshots taken previously become invalidated and shouldn't be used because meta backup and recovery don't replicate SST files. 

> Additionally, meta service maintains the retention time for SSTs required by meta snapshots via monitoring the snapshot storage in use to ensure consistency between meta snapshots and SST files. Thus, SST files required by meta snapshots from a snapshot storage that is not in use may be garbage collected at any time.

Reviewer 2: The previous "because" is mainly explained by this "Additionally" part. Currently it looks not so related to me.

------------------
                3
                4
```

## Delete a meta snapshot

Here's an example of how to delete a meta snapshot with `risectl`:

```
risectl meta delete-meta-snapshots [snapshot_ids]
```

## Restore from a meta snapshot

Use the following steps to restore from a meta snapshot.

1. Shut down the meta service.
    :::note
    This step is especially important because the meta backup and recovery process does not replicate SST files. It is not permitted for multiple clusters to run with the same SSTs set at any time, as this can corrupt the SST files.
    :::
2. Create an empty meta store.
3. Restore the meta snapshot to the new meta store.

    ```
    backup-restore \
    --meta-store-type etcd \
    --meta-snapshot-id [snapshot_id] \
    --etcd-endpoints [etcd_endpoints] \
    --storage-url [storage_url]
    ```
    `backup-restore` reads snapshot data from snapshot storage and writes them to `etcd`. 
    `backup-restore` is not included in the pre-built risingwave binary. Please build it from source by compiling the `risingwave_backup_cmd` package.
4. Configure meta service to use the new meta store.

## Access historical data backed up by meta snapshot

> Sometimes, a meta snapshot is used to support historical data access, known as a time travel query.

Reviewer 2: ```suggestion
Meta snapshot is used to support historical data access, also known as time travel query.
```
